pxr/usd/sdf/assetPathResolver.cpp

```diff
@@ -223,7 +223,22 @@ Sdf_ComputeAnonLayerIdentifier(
     const SdfLayer* layer)
 {
     TF_VERIFY(layer);
-    return TfStringPrintf(identifierTemplate.c_str(), layer);
+
+    // Identifier template is either "anon:" or "anon:SOME_TAG"
+    // and we respectively return "anon:LAYER_ADDRESS" or "anon:LAYER_ADDRESS:SOME_TAG".
+    // We use stringstream instead of TfStringPrintf so that URL-encoded characters such as
+    // %20 do not get misinterpreted as format strings, which triggers crashes in the
+    // string library.
+    std::stringstream strstream;
```

Comment:
std::stringstream brings in further behaviors, and can be slow compared to plain old concatenation. Since the form of the template is known, and no locale info and so on should be considered, using TfStringJoin here instead would avoid any new unexpected behaviors and be more performant than either stringstream or TfStringPrintf to boot.

Reply:
Thanks @meshula ! Will rework accordingly.

Reply:
Hi @meshula -- dumb question-- without TfStringPrintf or stringstream, how would we get the layer address into the string via TfStringJoin?

Secondly, SdfLayer::SdfLayer has

`    // If the identifier has the anonymous layer identifier prefix, it is a
    // template into which the layer address must be inserted. This ensures
    // that anonymous layers have unique identifiers, and can be referenced by
    // Sd object reprs.
    string layerIdentifier = Sdf_IsAnonLayerIdentifier(identifier) ?
        Sdf_ComputeAnonLayerIdentifier(identifier, this) : identifier;`
        
whereas testSdfLayer has 

`    def test_FindOrOpenNoAssetAnonLayers(self):
        """
        This test case confirms assumptions about how anonymous layers behave 
        for fileformats that can read/generate layers without requiring a 
        resolved asset to read. These assumptions allow anonymous layer 
        identifiers along with file format arguments to be used to refer to 
        fully dynamically generated layers without requiring a placeholder real 
        asset file that is never read.
        """`

which deliberately does not want the layer address inserted into the identifier for those cases. I'm not sure how best to manage the handshaking here between Sdf_GetAnonLayerIdentifierTemplate and Sdf_ComputeAnonLayerIdentifier, as it's possible for an input to come in with something like ":%p:" inside of a tag to Sdf_GetAnonLayerIdentifierTemplate  .

In the test_FindOrOpenNoAssetAnonLayers cases, Sdf_GetAnonLayerIdentifierTemplate is not called, so maybe it's possible to have Sdf_GetAnonLayerIdentifierTemplate prefix its return values with some sentinel other than "anon:", which Sdf_GetAnonLayerIdentifierTemplate then interprets as requiring the layer address in the identifier. But I'm not sure that's the most elegant and clear approach.

Thanks for reading!

Reply:
Hi Aaron, thanks for your patience!

We followed the rabbit down the hole on the `:%p:` business you note above!  Ultimately, just before  `SdfComputeAnonLayerIdentifier` is invoked, this routine is called, to poke in the layer identifier:

```cpp
string
Sdf_GetAnonLayerIdentifierTemplate(
    const string& tag)
{
    string idTag = tag.empty() ? tag : TfStringTrim(tag);
    return _Tokens->AnonLayerPrefix.GetString() + "%p" +
        (idTag.empty() ? idTag : ":" + idTag);
}
```

We think the most robust way to handle the issue you report, and preserve the important `:%p`  behavior, is to modify the `Sdf_GetAnonLayerIdentifierTemplate` routine so that `string idTag` is computed not by white space trimming tag, but by white space trimming the tag AND also doubling any percents in the input tag in order to safely escape them. (By "robust", here I mean that your URL-containing strings will be bottlenecked reliably through `Sdf_GetAnon...` making it a better fix site than `Sdf_ComputeAnon...`.) 

Side note: Ideally, a single loop might construct `idTag` by simultaneously trimming white space and duplicating the percent signs, but given the amount of string copying already exhibited in the current implementation of `Sdf_GetAnonLayerIdentifierTemplate`, I don't think it would make a difference to overly optimize it.

Reply:
Thanks @meshula ! I'll take a pass at it.